src/3d/qgs3dsceneexporter.cpp

```diff
@@ -0,0 +1,524 @@
+      for ( QgsChunkNode *node : leafs )
+      {
+        terrainTile = getDemTerrainEntity( terrain, node );
+        this->parseDemTile( terrainTile, textureGenerator );
+      }
+      break;
+    case QgsTerrainGenerator::Flat:
+      for ( QgsChunkNode *node : leafs )
+      {
+        terrainTile = getFlatTerrainEntity( terrain, node );
+        this->parseFlatTile( terrainTile, textureGenerator );
+      }
+      break;
+    // TODO: implement other terrain types
+    case QgsTerrainGenerator::Mesh:
+      break;
+    case QgsTerrainGenerator::Online:
+      break;
+  }
+  textureGenerator->setTextureSize( oldResolution );
+}
+
+QgsTerrainTileEntity *Qgs3DSceneExporter::getFlatTerrainEntity( QgsTerrainEntity *terrain, QgsChunkNode *node )
+{
+  QgsFlatTerrainGenerator *generator = dynamic_cast<QgsFlatTerrainGenerator *>( terrain->map3D().terrainGenerator() );
+  QgsChunkLoader *loader = generator->createSynchronousChunkLoader( node );
+  FlatTerrainChunkLoader *flatTerrainLoader = qobject_cast<FlatTerrainChunkLoader *>( loader );
+  if ( flatTerrainLoader == nullptr ) return nullptr;
+  // the entity we created should be deallocated when we deallocate the scene exporter
+  Qt3DCore::QEntity *entity = flatTerrainLoader->createEntity( this );
+  QgsTerrainTileEntity *tileEntity = qobject_cast<QgsTerrainTileEntity *>( entity );
+  return tileEntity;
+}
+
+QgsTerrainTileEntity *Qgs3DSceneExporter::getDemTerrainEntity( QgsTerrainEntity *terrain, QgsChunkNode *node )
+{
+  QgsTerrainTileEntity *tileEntity = nullptr;
+  // Just create a new tile (we don't need to export exact level of details as in the scene)
+  // create the entity synchronously and then it will be deleted once our scene exporter instance is deallocated
+  const Qgs3DMapSettings &map = terrain->map3D();
+  QgsDemTerrainGenerator *generator = static_cast<QgsDemTerrainGenerator *>( map.terrainGenerator() );
+  int oldResolution = generator->resolution();
+  generator->setResolution( mTerrainResolution );
+  QgsDemTerrainTileLoader *loader = qobject_cast<QgsDemTerrainTileLoader *>( generator->createSynchronousChunkLoader( node ) );
+  tileEntity = qobject_cast<QgsTerrainTileEntity *>( loader->createEntity( this ) );
+  generator->setResolution( oldResolution );
+  return tileEntity;
+}
+
+void Qgs3DSceneExporter::parseFlatTile( QgsTerrainTileEntity *tileEntity, QgsTerrainTextureGenerator *textureGenerator )
+{
+  Qt3DRender::QGeometryRenderer *mesh = nullptr;
+  Qt3DCore::QTransform *transform = nullptr;
+  for ( Qt3DCore::QComponent *component : tileEntity->components() )
+  {
+    Qt3DRender::QGeometryRenderer *meshTyped = qobject_cast<Qt3DRender::QGeometryRenderer *>( component );
+    if ( meshTyped != nullptr )
+    {
+      mesh = meshTyped;
+      continue;
+    }
+    Qt3DCore::QTransform *transformTyped = qobject_cast<Qt3DCore::QTransform *>( component );
+    if ( transformTyped != nullptr )
+    {
+      transform = transformTyped;
+    }
+  }
```

Comment:
It would be good to have some utility functions e.g. findGeometryRendererComponent(), findTransformComponent() to make code shorter here and in parseDemTile()...

Reply:
I made them into one function template named findTypedComponent